### PR TITLE
WIP - feat(popover): Add container input, support 'body' selector

### LINKED ISSUE #852: appendToBody option for Popover
It looks like the positioning service is ready to accept an `appendToBody` parameter, but this option isn't included in the NgbPopoverConfig. I took a crack at editing the transpiled javascript in the npm module to see if it was possible and sort of got it working, but I'm wondering if there's a known issue standing in the way of it getting implemented.

### PR DESCRIPTION
**Pull request by spongessuck**
Adds an `appendToRoot` input to the popover component to allow for placement of the popover in the root element. Addresses #852.

I'm not sure about the cross-platform implications of this solution, but it works in the browser, is analogous to the ui-bootstrap `appendToBody` option, and makes popover DOM placement more flexible until #339 is implemented.

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@spongessuck thnx for the PR, it looks like a first good step. Having said this I had a slightly different plan for this. My intention was to introduce `@Input` with the name `container` and type `string` and default value being undefined. For now this input would only kick in for the value equal to `body`. In the future we could extend / open up it to support any selector / element reference.

With my proposal we would be not introducing APIs that would be redundant / deprecated in the future.

Would you be willing to do the above changes in your PR?

**Comment by spongessuck** [comment]
Sure, it shouldn't be hard.

**Comment by spongessuck** [comment]
What about `container="root"`, to be consistent? The positioning service measures the offset based on the `html` element.

**Comment by pkozlowski-opensource** [comment]
> What about container="root", to be consistent? The positioning service measures the offset based on the html element.

I would prefer '"body"' for now as it will still work when we add support for all selectors.  

> Sure, it shouldn't be hard.

Awesome!

**Comment by spongessuck** [comment]
I'll do "html" then. I still think it's a bad idea to call it "body" when it's not measuring from the body.

**Comment by pkozlowski-opensource** [comment]
> I'll do "html" then. I still think it's a bad idea to call it "body" when it's not measuring from the body

Could we please stick to `body`? I understand your arguments but I want to keep `body` as default as this is what is shown in BS docs. And as things stand today it doesn't really matter what is measured from but rather where the element is attached.

**Comment by spongessuck** [comment]
If the body has any padding or margin I don't think it will be placed properly if it's in the body.

**Comment by spongessuck** [comment]
I guess it'd have to be positioned `relative` too... I'll do some tests, but do you see my concern?

**Comment by spongessuck** [comment]
Did some tests and it doesn't seem to matter where the body is- the position service is smarter than I thought :). I'll use "body" as requested.

File: src/popover/popover.ts [hunk 1]
```diff
@@ -95,6 +109,7 @@ export class NgbPopover implements OnInit, OnDestroy {
       this._windowRef = this._popupService.open(this.ngbPopover);
       this._windowRef.instance.placement = this.placement;
       this._windowRef.instance.title = this.popoverTitle;
+      this._windowRef.instance.container = this.container;
```
Comment (pkozlowski-opensource) [inline src/popover/popover.ts]: Does popover window need to be aware of the `container` option?

File: src/popover/popover.ts [hunk 2]
```diff
@@ -34,6 +34,7 @@ import {NgbPopoverConfig} from './popover-config';
 export class NgbPopoverWindow {
   @Input() placement: 'top' | 'bottom' | 'left' | 'right' = 'top';
   @Input() title: string;
+  @Input() container: string;
```
Comment (pkozlowski-opensource) [inline src/popover/popover.ts]: Why do we need this input?

File: src/popover/popover.spec.ts [hunk 3]
```diff
@@ -382,6 +403,7 @@ describe('ngb-popover', () => {
       config = c;
       config.placement = 'bottom';
       config.triggers = 'hover';
+      config.container = 'body';
```
Comment (pkozlowski-opensource) [inline src/popover/popover.spec.ts]: Why do we need to modify this test?

File: src/popover/popover.ts [hunk 4]
```diff
@@ -77,12 +83,20 @@ export class NgbPopover implements OnInit, OnDestroy {
...
 
     this._zoneSubscription = ngZone.onStable.subscribe(() => {
       if (this._windowRef) {
-        positionElements(this._elementRef.nativeElement, this._windowRef.location.nativeElement, this.placement);
+        positionElements(
+            this._elementRef.nativeElement, this._windowRef.location.nativeElement, this.placement,
+            this.container === 'body');
+
+        if (this.container === 'body') {
+          let windowEl = this._windowRef.location.nativeElement;
```
Comment (pkozlowski-opensource) [inline src/popover/popover.ts]: Nit: maybe drop the temporary `windowEl` variable?

File: src/popover/popover.spec.ts [hunk 5]
```diff
@@ -191,6 +193,25 @@ describe('ngb-popover', () => {
     });
   });
 
+  describe('placement', () => {
```
Comment (pkozlowski-opensource) [inline src/popover/popover.spec.ts]: Nit: 'placement' => 'container'?

**Comment by pkozlowski-opensource** [comment]
Thnx for all the updates @spongessuck - it all starts to come together 👍 
I've just left some minor comments that we need to sort out before merging.

File: src/popover/popover.ts (hunk 1 above)
Comment (spongessuck) [inline src/popover/popover.ts]: I guess it doesn't.

File: src/popover/popover.ts (hunk 2 above)
Comment (spongessuck) [inline src/popover/popover.ts]: Ditto.

File: src/popover/popover.spec.ts (hunk 3 above)
Comment (spongessuck) [inline src/popover/popover.spec.ts]: Doesn't this apply to popover and not popoverwindow? Shouldn't it test that it takes custom config properly?

File: src/popover/popover.spec.ts (hunk 5 above)
Comment (spongessuck) [inline src/popover/popover.spec.ts]: Sure.

File: src/popover/popover.ts (hunk 4 above)
Comment (pkozlowski-opensource) [inline src/popover/popover.ts]: Whatever you prefer, I'm good either way.

**Comment by spongessuck** [comment]
Not sure if you got notifs from my commits, but I think this is good to go.

**Comment by pkozlowski-opensource** [comment]
Thnx @spongessuck - squashed and merged as d1a22c0. Would be awesome if you could send a similar (well, almost identical) PR for tooltips :-)

**Pull request closed without merging**